### ISSUE TITLE
Support Infrastructure operations running for longer than 60 min with temporary credentials

### ISSUE THREAD

**Issue by mxro** [goldstack/goldstack#99]
Changes introduced in [v0.3.15](https://github.com/goldstack/goldstack/releases/tag/v0.3.15) allow supplying temporary (session) AWS credentials for infrastructure operations. Many of these expire in 60 min. 

This would lead to problems in the current implementation with infrastructure operations that run for longer than 60 min, since there is no mechanims for refreshing the credentials https://github.com/goldstack/goldstack/issues/17#issuecomment-1045041531

Possilbe solutions:

- Mount a directory that contains AWS config and/or credentials file into the Docker container and provide a local path to this file as environment variables e.g. `AWS_CONFIG_FILE`

**Comment by chrismilleruk** [comment]
Looking at potential solutions from best to worst, I think it looks like this:

1. get terraform to refresh the credentials if they expire during a long-running operation
1. check expiry before you launch a terraform command and 
    1. manually refresh if there is < x minutes remaining.
    1. terminate the container (gracefully?) if it runs over.
1. check expiry when you first retrieve the credentials and do something (i.e. warn the user)

TLDR; No magic fixes unfortunately but a couple of potential workarounds.

### 1. get terraform to refresh the credentials if they expire during a long-running operation
> https://registry.terraform.io/providers/hashicorp/aws/latest/docs
> The following methods are supported, in this order, and explained below:
> * Static credentials
> * Profiles with shared credentials and configuration files
> * Environment variables
> * CodeBuild, ECS, and EKS Roles
> * EC2 Instance Metadata Service (IMDS and IMDSv2)

`Static credentials` and `Environment variables` can't be refreshed.
`CodeBuild, ECS, and EKS Roles` and `EC2 Instance Metadata Service` can't be faked up so far as I can tell. They expect an HTTP service at `169.254.169.254` or `169.254.0.3` and while it might be possible to achieve this using docker-compose and some network trickery, it feels overkill for the problem.
`Profiles with shared credentials and configuration files` *should* be able to refresh using `credential_process` (assuming AWS Go SDK supports it) *BUT* the container doesn't have access to the host `credential_process` cmd, AND this would need to be executed on the host (where the `AWS SSO login` took place)

**If `terraform` was run on the host instead of a container (i.e. using `tfenv` to switch versions), then the `profile` approach *might work* because it exists in the same security context.**

## 2. check expiry before you launch a terraform command.
I couldn't see a way to reliably get the expiry after initial auth so we would have to store it somewhere and refer to the initially supplied value. If we're doing that, we could also store a reference to the `refreshPromise()` function.

### i. manually refresh if there is < x minutes remaining.
`refreshPromise()` will provide new credentials if there is less than `expiryWindow` seconds remaining. This defaults to 15 seconds so would need to be changed. I did some lightweight testing with this but couldn't get it working. Have just noticed now that it's a static property, so `AWS.Credentials.expiryWindow = 15` might work better (albeit with side effect).

Moving past that, what value would you want to set it to? Most terraform commands take only seconds to complete, so it's just one or two key cases that we should focus on.

### ii. terminate the container (gracefully?) if it runs over.

Not sure how to do this gracefully, also I found that when I terminated a long running terraform command before it caused a lock-state which was difficult to undo via `yarn infra xx`. Created #101 for this.

## 3. check expiry when you first retrieve the credentials and do something (i.e. warn the user)

This might actually be the simplest approach. We get an expiryTime from the `credential_process` response and if it's 60 minutes or less that we can warn the user and allow them to abort.

SSO credential expiries range from 1 hour to 12 hours. This are all configurable by the SSO administrator but we should assume they can't be altered. I recently set up SSO using the defaults and I noticed that AdministratorAccess seems to get a 1 hour expiry whereas PowerUser gets 8 hours. 

We're likely to breach 60 minutes if we're setting up everything for the first time and/or if DNS infra needs changing. 

I *think* we only need AdministratorAccess to create users, once we're past that, PowerUser should suffice (needs testing).

It feels like it might be possible to use the Admin profile for the basics and put all the long-running stuff through a PowerUser profile (which is probably better in any case).

**Comment by mxro** [comment]
From observations so far, I think that most operations to create infrastructure seem to finish within 20-30 min (even involving CF and ACM). But `destroy` operations may break it.

See here workflow runs that will plan, apply & destroy the infrastructure in the templates (plus do a deploy): https://github.com/goldstack/goldstack/runs/5262006788?check_suite_focus=true - which seem to happen in under 30 min for the most part for all the templates.

Could there be any update operations that take longer than this? The idea here that I had (not 100% sure if it is a good one) was to break up larger project by allowing each package to have its own TF, thereby hopefully preventing very long TF applies - but that approach of course brings different problems!
